render duration of intro & outro ranges on overview waveforms

Be-ing opened the pull request:
MarkRange elements in skins only have their durations rendered if they specify a DurationTextColor. This avoids rendering the durations of loops on the overview waveform.

![image](https://user-images.githubusercontent.com/9455094/56545278-b6c02680-653c-11e9-9636-0f753728c057.png)

durations scale with rate slider:
![image](https://user-images.githubusercontent.com/9455094/56545312-d48d8b80-653c-11e9-8d91-1a0034f7f78e.png)

I played a 3 hour set with this last night and it is extremely helpful for me. Having Mixxx calculate these times for me allows me to be more intentional about how I align tracks. I used to do this in headphones continually. With this I don't have to keep the times in my head, I don't have to do any math, and I don't have to repeat that the next time I play a track.

I understand this could be distracting or clutter for others, so I'll add an option to disable this unless there are objections to adding an option.

Be-ing committed:
render duration of intro & outro ranges on overview waveforms

MarkRange elements in skins only have their durations rendered if they
specify a DurationTextColor. This avoids rendering the durations of
loops on the overview waveform.

uklotzde commented on src/widget/woverview.cpp:
```diff
@@ -565,6 +533,65 @@ void WOverview::paintEvent(QPaintEvent * /*unused*/) {
+                    // TODO: replace with rate_ratio in PR #1765
+                    double rateRatio = 1.0 + m_pRateDirControl->get() * m_pRateRangeControl->get() * m_pRateSliderControl->get();
+                    QString duration = mixxx::Duration::formatTime((endValue - startValue)
+                            / m_trackSampleRateControl->get() / 2 / rateRatio);
```
It might be worth to define a constant for the number of channels to not forget about it once the engine is rewritten.

uklotzde commented:
Very helpful feature :+1:

Instead of only enabling/disabling the display we could think about specifying a minimum duration, e.g. with a configurable range between 0 to 30 seconds and a default value of 3 or 5 seconds. This value may vary depending on your music and mixing style. Not sure if this is more confusing than helpful?

Now I also noticed that I need markers on the overview waveforms for a quick peek how long a phrase might be :see_no_evil:

Be-ing commented:
> Not sure if this is more confusing than helpful?

I think that would be more confusing than helpful.

daschuer commented:
How about this appearance: 
![grafik](https://user-images.githubusercontent.com/1777442/57008989-2c3e8d80-6bf4-11e9-90b1-344dac5c364c.png)

We can move the time label conditionally into the region if enough space is available. 

I don't think we need a preferences option for it.

Be-ing commented:
Why do you want the text moved outside the range? And why move it to the top?

If we can find a solution that everyone is satisfied with that does not require a preference option that would be ideal.

daschuer commented:
It look IMHO ugly if the time labele escapes the boundaries of the highlighted region. 
Like proposed it labels the triangle "00:15 ◣"  and "◢ 00:23" which keeps all the eye candy in one place.

Be-ing commented:
To clarify, are you proposing putting the text outside the range on the side closer to the middle of the track? So for the intro it would be drawn after the range and for the outro it would be drawn before the range?

uklotzde commented:
This is exactly how I understood Daniel's proposal. The labels would be displayed outside of the (probably short region) within the main track range, next to their nearest marker.

For short intro/outro ranges the labels might not fit into their sections and would otherwise be displayed across the boundaries with varying backgrounds.

There is no ideal solution. With Daniel's proposal the appearance is slightly more consistent.

uklotzde commented:
A typical example of a short intro:
![Screenshot from 2019-05-01 20-41-35](https://user-images.githubusercontent.com/1474154/57035207-be28b380-6c51-11e9-93be-af2aa232d227.png)

Reusing the translucent background from highlighted region for the label might be an option to improve readability. The waveform will still shine through.

daschuer commented:
> To clarify, are you proposing putting the text outside the range on the side closer to the middle of the track? So for the intro it would be drawn after the range and for the outro it would be drawn before the range?

Yes, as an idea for discussion. The part I like most is the idea to have the texts and the triangle close to each other and not draw across the region borders.

Be-ing committed:
move intro/outro duration text outside of range on overviews

This avoids the text overlapping the borders of the range, which
looks ugly.

Be-ing committed:
fix Deere outro start position location on overview

to match all the other skins

Be-ing committed:
truncate leading 0 from duration strings

This saves a little space, which is helpful for the duration
text for intro/outro sections on the overview waveforms.

Be-ing commented:
How is this? Do we still need to add a preference option?
![image](https://user-images.githubusercontent.com/9455094/57197211-2ca79300-6f2a-11e9-8c50-ba2bc8c8d029.png)

daschuer commented:
Ok, thank you. 
Is the ◣ under the blue overlay? It should have the same color than the time text.

Be-ing committed:
draw cue markers on top of ranges on overview

This was changed in 5161968 so the cue markers did not get drawn
over range duration text, but that concern was obviated by
0ab6a46

Be-ing commented:
Good catch, fixed:
![image](https://user-images.githubusercontent.com/9455094/57197357-f408b900-6f2b-11e9-952f-c428db8ddcd3.png)

daschuer commented:
LGTM Thanks, waiting for CI.

Be-ing committed:
update DurationUtilTest formatTime

Be-ing committed:
use mixxx::kEngineChannelCount constant instead of '2' literal

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into intro_outro_time

Be-ing commented:
CI passed :+1:

Be-ing commented:
If there is still a demand for adding an option to disable this, that can still be added during the 2.3 beta period. Hopefully this looks good enough that it isn't bothersome enough to anyone that an option is required.
